mountpoint-s3/src/store.rs

```diff
@@ -0,0 +1,235 @@
+use std::fmt::Debug;
+
+use async_trait::async_trait;
+use futures::{executor::ThreadPool, task::Spawn};
+use mountpoint_s3_client::error::{
+    DeleteObjectError, GetObjectAttributesError, GetObjectError, HeadObjectError, ListObjectsError, ObjectClientError,
+    PutObjectError,
+};
+use mountpoint_s3_client::types::{
+    DeleteObjectResult, ETag, GetObjectAttributesResult, HeadObjectResult, ListObjectsResult, ObjectAttribute,
+    ObjectClientResult, PutObjectParams,
+};
+use mountpoint_s3_client::{ObjectClient, PutObjectRequest};
+use thiserror::Error;
+
+use crate::checksums::{ChecksummedBytes, IntegrityError};
+use crate::prefetch::{self, ClientPartStream, ObjectPartStream, Prefetcher, PrefetcherConfig};
+use crate::sync::Arc;
+
+/// A generic interface to S3-like object store.
+/// Similar to [ObjectClient], but provides a [ObjectStore::prefetch] method instead
+/// of [ObjectClient::get_object].
+#[async_trait]
+pub trait ObjectStore: Clone {
```

Comment:
See discussion on extending `ObjectClient`: https://github.com/awslabs/mountpoint-s3/pull/590#discussion_r1378813014

Reply:
I think extending `ObjectClient`, or otherwise exposing the `ObjectClient` trait directly, is still the right thing to do here. 

Actually, thinking more broadly: right now this is a pretty heavyweight abstraction for what's essentially one method (`prefetch`) with two implementations (cached or not). Seems like we could have a narrower interface that just focuses on "where object bytes come from". For example, I don't see how we'd use this abstraction to handle cached writes in the future.

Basically: can we just start with a smaller abstraction and expand it once we need it?

Reply:
An alternative approach could be to introduce a `Prefetcher` trait, roughly something like:

```rust
trait Prefetcher {
  type ClientError: std::error::Error + Send + Sync + 'static;
  type PrefetchGetObject: PrefetchGetObject<ClientError = Self::ClientError>;
  fn prefetch(&self, bucket: &str, key: &str, size: u64, etag: ETag) -> Self::PrefetchGetObject;
}
```

Comparing with `ObjectStore`:

Pros:
* focus on "where object bytes come from"
* no impact on write path, readdir, etc.

Cons:
* the prefetcher type would still need to be propagated together with the client in many places, e.g.:
  `struct S3Filesystem<Client, Prefetcher>`.
* If we wanted to introduce cached writes in the future, we would have to carry even more types, e.g.: `struct S3Filesystem<Client, Prefetcher, Uploader>`.
* building a filesystem may become a bit complex and error prone, e.g.:
  ```rust
  let client = Arc::new(client); // Annoyingly, ObjectClient is not Clone
  let prefetcher = SomePrefetcher::new(client.clone());
  let filesystem = S3Filesystem::new(client, ...);
  ```
  Although maybe solvable it we had `Prefetcher::prefetch` take `Client` as an argument (or rather `Arc<Client>`). 

On balance, I would still lean towards `ObjectStore`.

Reply:
Another idea could be something like:

```rust
trait ObjectStore {
  ...
  type Client: ObjectClient + ...
  
  fn client(&self) -> &ObjectClient;

  fn prefetch(...) -> ...

  // in the future 
  fn write(...)
}
```

Not too sure if it would qualify as a smaller abstraction, but it would make it clear what is the relationship with `ObjectClient`.

Reply:
I don't think "too many type parameters" is much of a concern — it's annoying to get right one time, and then shouldn't really matter. I like the idea of having small, separate abstractions for this until we know for sure that we need a shared one (for e.g. reads and writes).

Reply:
I'll start reworking with something like the `Prefetcher` above. When I'm done, shall I update this PR, or open a new one?

Reply:
Opened #595